File: typescript/src/runtime/marking.ts
/**
 * @module marking
 *
 * Mutable token state (marking) of a running Petri net.
 *
 * Tokens per place are stored in FIFO arrays (push to end, shift from front).
 * Not thread-safe — all mutation occurs from the single-threaded orchestrator.
 *
 * For typical nets (≤10 tokens per place), Array.shift() is faster than a deque
 * due to cache locality. Only optimize if profiling shows a bottleneck.
 */
import type { Place } from '../core/place.js';
import type { Token } from '../core/token.js';
import { compareCodePoints } from '../core/internal/code-point-order.js';

const EMPTY_TOKENS: readonly Token<any>[] = Object.freeze([]);

/**
 * A place reference plus an optional per-token predicate.
 *
 * Used by the ν-net join path (NU-020/NU-021) to select the tokens whose
 * projected correlation name equals the chosen binding. This is *not* an input
 * guard (IO-006, removed) — input specifications are purely structural; the
 * predicate here is always derived from name correlation.
 */
export interface PredicateSpec<T = any> {
  readonly place: Place<T>;
  readonly predicate?: (value: T) => boolean;
}

/**
 * A marking captured in the normative snapshot form of **CORE-073**: place **name** to that
 * place's FIFO token sequence.
 *
 * The entry type is the ordinary {@link Token} — it already *is* `(value, createdAt)`, and a
 * parallel `SnapshotEntry` would be a second name for one shape, which is the duplication
 * CORE-073 exists to prevent.
 *
 * A `Map` rather than a plain object, for insertion order and because a place named
 * `__proto__` or `constructor` would otherwise collide with a prototype key. That is reachable
 * input rather than a hypothetical: a host compiling place names from user-supplied step or
 * node identifiers can be handed either. Java (`Map<String, List<Token<?>>>`) and Rust
 * (`BTreeMap<Arc<str>, Vec<ErasedToken>>`) carry the structurally identical form.
 *
 * Named `…Form` rather than `MarkingSnapshot` because that name is already taken by the
 * `marking-snapshot` **event** payload — the third instance of this collision across the three
 * languages: Java's is with the same event record, Rust's is with `ExecutorSignal::Snapshot`.
 * The name is TypeScript-local: Java carries the identical structure as a raw map type with no
 * alias, and Rust names it `MarkingSnapshot` (`pub type`, re-exported from the crate root). An
 * alias is a name for the structure, not a second structure, so none of this forks the agreed
 * form.
 *
 * @see Marking.snapshot
 * @see Marking.fromSnapshot
 */
export type MarkingSnapshotForm = ReadonlyMap<string, readonly Token<unknown>[]>;

/**
 * Mutable marking (token state) of a Petri Net during execution.
 *
 * Tokens in each place are maintained in FIFO order (array: push to end, shift from front).
 * Not thread-safe — all access must be from the orchestrator.
 */
export class Marking {
  /** Place name -> FIFO queue of tokens. */
  private readonly tokens = new Map<string, Token<any>[]>();
  /** Place name -> Place reference (for inspection). */
  private readonly placeRefs = new Map<string, Place<any>>();

  static empty(): Marking {
    return new Marking();
  }

  /**
   * Captures this marking in the normative snapshot form (**CORE-073**): place **name** to
   * that place's FIFO token sequence.
   *
   * Keyed by name rather than by {@link Place} because a snapshot MUST survive a round-trip
   * through a net that has since gained or lost a place (CORE-073 AC#7, [CORE-072]) — there
   * is no `Place` to construct for a name the receiving net never declared.
   *
   * **Structural, not serialized.** Token values are carried through untouched; the engine
   * imposes no codec and does not require them to be serializable, because only the host
   * knows what its values are. A closure, a class instance or a native handle round-trips
   * within the process unchanged (AC#8). Persisting a snapshot across processes is therefore
   * the host's encoding to choose.
   *
   * Each sequence is an owned copy in FIFO order ([CORE-013]); token order decides which
   * token a firing consumes, so a snapshot that lost it would not be a restore (AC#6).
   * Places are emitted in code-point order of their names. The form is a mapping, so only the
   * per-place sequence is *required* to be ordered — but a deterministic key order makes a
   * snapshot content-addressable, makes two captures diffable without normalising first, and
   * matches the order Rust's `BTreeMap<Arc<str>, _>` and Java's `TreeMap` produce, so two
   * implementations snapshotting one marking agree key for key (AC#12). Do not relax it.
   *
   * **Empty places are omitted.** A place that was drained, or restored empty, does not appear.
   * One emission rule is what makes "agree key for key" achievable at all: Java, Rust and the
   * precompiled executor never had such keys to emit, and a `Marking` cannot tell "declared and
   * drained" from "never there" anyway — it only knows places that once held a token.
   * {@link fromSnapshot} still accepts both forms and restores them identically (AC#7).
   *
   * **Persisting one.** The values are yours. If they are not JSON-safe that is a host concern
   * this engine cannot answer — a codec hook would not help, since "what does this closure mean
   * in another process" is a question only the host can answer. The walk is a line each way, as
   * an **array of entries**: `JSON.stringify([...marking.snapshot()])` on the way to storage,
   * and `new Map(JSON.parse(stored))` on the way back into {@link fromSnapshot}.
   *
   * Not `Object.fromEntries(...)`. A JSON *object* is an unordered medium, and a JavaScript
   * object is worse than unordered: integer-like keys are hoisted ahead of every other key in
   * numeric order, so places named `'9'`, `'10'`, `'-1'`, `'a'` — canonical order
   * `-1, 10, 9, a` — come back as `9, 10, -1, a`. A host naming places after node ids hits
   * that at once. It still *restores* correctly, because a restore ignores the order it is
   * handed; what is lost is the canonical order of the stored artefact, and with it the
   * byte-comparability that order exists for. An array of entries keeps it, and needs no
   * reasoning about a place named `__proto__` either.
   */
  snapshot(): MarkingSnapshotForm {
    const names: string[] = [];
    for (const [name, queue] of this.tokens) {
      if (queue.length > 0) names.push(name);
    }
    names.sort(compareCodePoints);
    const out = new Map<string, readonly Token<unknown>[]>();
    for (const name of names) {
      out.set(name, [...this.tokens.get(name)!] as readonly Token<unknown>[]);
    }
    return out;
  }

  /**
   * Restores a marking from the snapshot form (**CORE-073**).
   *
   * Accepts a place that is **absent** and one **present but empty** as the same thing, per
   * AC#7 — an emitted snapshot omits empties, a hand-built or foreign one may carry them, and
   * this side carries the obligation to treat the two alike.
   *
   * Token `createdAt` values are carried through verbatim: the engine MUST NOT re-stamp a
   * restored token, including under an injected epoch clock ([TIME-015] AC#9 / CORE-073
   * AC#9). Restoring is the one path where the engine hands back a timestamp it did not
   * choose, and re-stamping would make a resume indistinguishable from a fresh start.
   *
   * The restored marking carries no {@link Place} references — a snapshot holds names only.
   * That is what lets a name the receiving net does not declare survive the round-trip
   * ([CORE-072]); such tokens are retained and inert.
   *
   * **The clocks do not come back.** A restored token's `createdAt` is metadata, not a firing
   * clock: every transition this marking enables starts its timing interval fresh at the
   * resuming executor's enablement ([TIME-010], [TIME-011]). A `Delayed` lower bound is still
   * satisfied — measured again from zero it is still at least that bound — but a `Deadline` or
   * `Window` **upper** bound receives a fresh full budget, so a deadline the net promised can
   * be silently missed across a restore. A hard bound that must survive one belongs in the
   * token payload or in an action timeout ([IO-013]), not in net timing.
   *
   * **Restore must stay occasional.** Every restore restarts every clock, so a transition
   * whose `Delayed(d)` clock is restarted more often than every `d` never accumulates `d` and
   * never fires — starvation, not slowdown. Sound for a suspend boundary or a checkpoint;
   * unsound for a scheduler that parks and resumes as a matter of course.
   */
  static fromSnapshot(snapshot: MarkingSnapshotForm): Marking {
    const m = new Marking();
    for (const [name, tokens] of snapshot) {
      if (tokens.length === 0) continue; // absent and empty are one thing (AC#7)
      m.tokens.set(name, [...tokens]);
    }
    return m;
  }

  static from(initial: Map<Place<any>, Token<any>[]>): Marking {
    const m = new Marking();
    for (const [place, tokens] of initial) {
      m.placeRefs.set(place.name, place);
      m.tokens.set(place.name, [...tokens]);
    }
    return m;
  }

  // ======================== Token Addition ========================

  addToken<T>(place: Place<T>, token: Token<T>): void {
    this.placeRefs.set(place.name, place);
    const queue = this.tokens.get(place.name);
    if (queue) {
      queue.push(token);
    } else {
      this.tokens.set(place.name, [token]);
    }
  }

  // ======================== Token Removal ========================

  /** Removes and returns the oldest token. Returns null if empty. */
  removeFirst<T>(place: Place<T>): Token<T> | null {
    const queue = this.tokens.get(place.name);
    if (!queue || queue.length === 0) return null;
    return queue.shift() as Token<T>;
  }

  /** Removes and returns all tokens from a place. */
  removeAll<T>(place: Place<T>): Token<T>[] {
    const queue = this.tokens.get(place.name);
    if (!queue || queue.length === 0) return [];
    const result = [...queue] as Token<T>[];
    queue.length = 0;
    return result;
  }

  /**
   * Removes and returns the first token whose value satisfies the predicate.
   *
   * Performs a linear scan of the place's FIFO queue. If no predicate is
   * provided, behaves like `removeFirst()`. Otherwise skips non-matching
   * tokens and splices the first match out of the queue (O(n) worst case).
   */
  removeFirstMatching(spec: PredicateSpec): Token<any> | null {
    const queue = this.tokens.get(spec.place.name);
    if (!queue || queue.length === 0) return null;
    if (!spec.predicate) {
      return queue.shift()!;
    }
    for (let i = 0; i < queue.length; i++) {
      const token = queue[i]!;
      if (spec.predicate(token.value)) {
        // Head removal (the common ν-net case — the matched token is the oldest,
        // hence at the front with distinct timestamps) uses the V8-optimized
        // shift() rather than an O(n) splice, keeping a draining join linear.
        if (i === 0) queue.shift();
        else queue.splice(i, 1);
        return token;
      }
    }
    return null;
  }

  // ======================== Token Inspection ========================

  /** Check if any token satisfies the predicate. */
  hasMatchingToken(spec: PredicateSpec): boolean {
    const queue = this.tokens.get(spec.place.name);
    if (!queue || queue.length === 0) return false;
    if (!spec.predicate) return true;
    return queue.some(t => spec.predicate!(t.value));
  }

  /**
   * Counts tokens in a place whose values satisfy the predicate.
   *
   * If no predicate is provided, returns the total token count (O(1)).
   * With a predicate, performs a linear scan over all tokens (O(n)).
   * Used by the executor to size a ν-net correlated `all`/`at-least` consume.
   */
  countMatching(spec: PredicateSpec): number {
    const queue = this.tokens.get(spec.place.name);
    if (!queue || queue.length === 0) return 0;
    if (!spec.predicate) return queue.length;
    let count = 0;
    for (const t of queue) {
      if (spec.predicate(t.value)) count++;
    }
    return count;
  }

  /**
   * Returns tokens in a place. **Returns a live reference** to the internal
   * array — callers must not mutate it. Copy with `[...peekTokens(p)]` if
   * mutation or snapshot semantics are needed.
   */
  peekTokens<T>(place: Place<T>): readonly Token<T>[] {
    return (this.tokens.get(place.name) ?? EMPTY_TOKENS) as readonly Token<T>[];
  }

  /** Returns the oldest token without removing it. */
  peekFirst<T>(place: Place<T>): Token<T> | null {
    const queue = this.tokens.get(place.name);
    return queue && queue.length > 0 ? queue[0] as Token<T> : null;
  }

  /** Checks if a place has any tokens. */
  hasTokens(place: Place<any>): boolean {
    const queue = this.tokens.get(place.name);
    return queue !== undefined && queue.length > 0;
  }

  /** Returns the number of tokens in a place. */
  tokenCount(place: Place<any>): number {
    const queue = this.tokens.get(place.name);
    return queue ? queue.length : 0;
  }

  // ======================== Debugging ========================

  toString(): string {
    const parts: string[] = [];
    for (const [name, queue] of this.tokens) {
      if (queue.length > 0) {
        parts.push(`${name}: ${queue.length}`);
      }
    }
    return `Marking{${parts.join(', ')}}`;
  }
}

File: rust/libpetri-runtime/src/marking.rs
use std::any::Any;
use std::collections::{BTreeMap, HashMap, VecDeque};
use std::sync::Arc;

use libpetri_core::place::Place;
use libpetri_core::token::{ErasedToken, Token};

/// \[CORE-073\] The normative snapshot form: place **name** → the place's
/// tokens in **FIFO order** (\[CORE-013\]).
///
/// Keyed by name rather than by [`Place`] deliberately: \[CORE-073\] AC#7
/// requires a snapshot to retain a place the receiving net does not declare,
/// and there is no `Place` to construct for a name the net has never heard
/// of. `Marking` has always been keyed this way, so retention is structural
/// rather than something to remember.
///
/// A `BTreeMap` rather than a `HashMap` so the place order is **deterministic
/// across runs**. The form is a mapping and only the per-place sequence is
/// ordered, so hash order would not violate the requirement — but the moment
/// a host serializes a snapshot and diffs, hashes or content-addresses it,
/// unstable key order bites, and ordering it costs nothing here.
///
/// The entry is [`ErasedToken`], which already *is* `(value, created_at)` —
/// no parallel "snapshot entry" type, matching Java and TypeScript, which
/// reuse their `Token` for the same reason. Note it carries one extra field,
/// `value_type_name`, which is engine metadata outside the normative form
/// and does not survive a round-trip through a binding.
///
/// **Structural, not serialized.** No codec is imposed and token values need
/// not be serializable: the value stays an `Arc<dyn Any + Send + Sync>`, so
/// a closure, a native handle or a host object round-trips in-process
/// untouched (\[CORE-073\] AC#8).
pub type MarkingSnapshot = BTreeMap<Arc<str>, Vec<ErasedToken>>;

/// \[ENV-014\] AC5/AC6 — what a mid-execution snapshot returns.
///
/// The marking **and** whether an action was in flight when it was taken,
/// as one value. Deliberately not a separately-queryable flag: that would be
/// read at a different instant than the marking, which is the race AC5
/// exists to close.
///
/// # Why the flag matters
///
/// Nothing suspends the run to serve a snapshot. An action may be **in
/// flight** — having already consumed its inputs (\[EXEC-031\]) and not yet
/// produced its outputs — so those tokens are in *neither* place at the
/// instant you observe them. The marking is a perfectly valid
/// **observation**; it is not a valid **restore point**, because restoring
/// it loses that work silently.
///
/// # External events
///
/// The flag's cross-language meaning is *work in flight: an action, or an
/// accepted but un-injected external event*. In this implementation the
/// second half is never true: an inject and a snapshot request travel the
/// same FIFO signal channel, and the executor applies an inject to the
/// marking the moment it receives it. Every event accepted before the
/// snapshot was requested is therefore already **in** the marking, and the
/// flag reduces to "an action was in flight".
///
/// `action_in_flight` is singular and boolean on purpose. A count would
/// invite the reader to treat it as a live gauge of how many are running
/// now, which is exactly the racy reading AC5 forbids. This is a fact about
/// the instant the snapshot was taken, not a measurement you can re-check.
#[derive(Debug, Clone)]
pub struct SnapshotResult {
    /// The captured marking, in the \[CORE-073\] normative form.
    pub marking: MarkingSnapshot,
    /// Work in flight: an action, or an accepted but un-injected external
    /// event (the latter cannot occur here — see *External events* above).
    /// True makes the marking an observation rather than a restore point.
    pub action_in_flight: bool,
}

impl SnapshotResult {
    /// True when nothing was in flight, so the marking is safe to restore
    /// from. Sugar for `!action_in_flight`, named for the question a caller
    /// is actually asking.
    pub fn is_restore_point(&self) -> bool {
        !self.action_in_flight
    }
}

/// Mutable token state of a Petri net during execution.
///
/// Stores type-erased tokens in FIFO queues keyed by place name. `Clone` is
/// supported so an executor can hand out an owned copy through a oneshot
/// channel (used by [`ExecutorHandle::snapshot`](crate::ExecutorHandle::snapshot));
/// each cloned token clones its underlying `Arc<dyn Any + Send + Sync>`, so
/// shared payloads are not deep-copied.
///
/// `Clone` is an **in-process copy, not the snapshot form** — see
/// [`snapshot`](Self::snapshot) / [`from_snapshot`](Self::from_snapshot) for
/// \[CORE-073\]. The two coexist; neither replaces the other.
#[derive(Debug, Default, Clone)]
pub struct Marking {
    tokens: HashMap<Arc<str>, VecDeque<ErasedToken>>,
}

impl Marking {
    pub fn new() -> Self {
        Self::default()
    }

    /// \[CORE-073\] Captures this marking in the normative snapshot form.
    ///
    /// Every token's `value` and `created_at` are preserved, and each
    /// place's sequence is its FIFO order, so a restore reproduces which
    /// token the next firing consumes (\[CORE-013\], AC#6).
    ///
    /// Empty places are omitted. \[CORE-073\] permits either, and requires
    /// only that omitted and present-but-empty restore identically —
    /// [`from_snapshot`](Self::from_snapshot) accepts both.
    ///
    /// Restoring does **not** resume a partially elapsed firing interval: a
    /// restored marking seeds a new execution whose timing clocks all start
    /// fresh (\[TIME-010\] / \[TIME-011\]). That is safe for `delayed` /
    /// `exact` lower bounds and **unsafe for `deadline` / `window` upper
    /// bounds**, which receive a fresh full budget — and restoring more often
    /// than a `Delayed(d)` interval means that transition never fires at all.
    /// Restore is sound as an *occasional* operation, not as a scheduler's
    /// routine park-and-resume.
    pub fn snapshot(&self) -> MarkingSnapshot {
        self.tokens
            .iter()
            .filter(|(_, queue)| !queue.is_empty())
            .map(|(place, queue)| (Arc::clone(place), queue.iter().cloned().collect()))
            .collect()
    }

    /// \[CORE-073\] Rebuilds a marking from the normative snapshot form.
    ///
    /// FIFO order within each place is preserved, `created_at` is carried
    /// through **unchanged** — the engine never re-stamps a restored token,
    /// including under an injected epoch clock (\[TIME-015\], AC#9) — and a
    /// place the receiving net does not declare is retained rather than
    /// dropped, exactly as \[CORE-072\] requires of any initial marking
    /// (AC#7).
    ///
    /// An explicitly-empty sequence is accepted and yields the same marking
    /// as omitting the place entirely (AC#7). That half is worth testing
    /// directly: a snapshot-then-restore round-trip can never exercise it,
    /// because [`snapshot`](Self::snapshot) never emits one.
    pub fn from_snapshot(snapshot: &MarkingSnapshot) -> Self {
        let mut marking = Self::new();
        for (place, entries) in snapshot {
            if entries.is_empty() {
                continue;
            }
            marking
                .tokens
                .entry(Arc::clone(place))
                .or_default()
                .extend(entries.iter().cloned());
        }
        marking
    }

    /// Adds a typed token to a place.
    pub fn add<T: Send + Sync + 'static>(&mut self, place: &Place<T>, token: Token<T>) {
        let erased = ErasedToken::from_typed(&token);
        self.tokens
            .entry(Arc::clone(place.name_arc()))
            .or_default()
            .push_back(erased);
    }

    /// Adds a type-erased token to a place by name.
    pub fn add_erased(&mut self, place_name: &Arc<str>, token: ErasedToken) {
        self.tokens
            .entry(Arc::clone(place_name))
            .or_default()
            .push_back(token);
    }

    /// Returns the number of tokens in a place.
    pub fn count(&self, place_name: &str) -> usize {
        self.tokens.get(place_name).map_or(0, |q| q.len())
    }

    /// Returns true if a place has any tokens.
    pub fn has_tokens(&self, place_name: &str) -> bool {
        self.count(place_name) > 0
    }

    /// Peeks at the first token value in a place without removing it.
    pub fn peek<T: Send + Sync + 'static>(&self, place: &Place<T>) -> Option<Arc<T>> {
        self.tokens
            .get(place.name())
            .and_then(|q| q.front())
            .and_then(|t| t.downcast::<T>().map(|token| token.value_arc()))
    }

    /// Removes and returns the first token from a place (FIFO).
    pub fn remove_first(&mut self, place_name: &str) -> Option<ErasedToken> {
        self.tokens.get_mut(place_name).and_then(|q| q.pop_front())
    }

    /// Removes and returns the first token whose value satisfies `predicate`.
    ///
    /// Used by the ν-net join path (NU-020/NU-021) to take the tokens whose
    /// projected correlation name equals the chosen binding. This is *not* an
    /// input guard (IO-006, removed) — input specifications are purely
    /// structural; the predicate here is always derived from name correlation.
    pub fn remove_matching(
        &mut self,
        place_name: &str,
        predicate: &dyn Fn(&dyn Any) -> bool,
    ) -> Option<ErasedToken> {
        let queue = self.tokens.get_mut(place_name)?;
        let pos = queue.iter().position(|t| predicate(t.value.as_ref()))?;
        queue.remove(pos)
    }

    /// Removes and returns all tokens from a place.
    pub fn remove_all(&mut self, place_name: &str) -> Vec<ErasedToken> {
        self.tokens
            .get_mut(place_name)
            .map_or_else(Vec::new, |q| q.drain(..).collect())
    }

    /// Removes and returns all tokens whose values satisfy `predicate`.
    pub fn remove_all_matching(
        &mut self,
        place_name: &str,
        predicate: &dyn Fn(&dyn Any) -> bool,
    ) -> Vec<ErasedToken> {
        let queue = match self.tokens.get_mut(place_name) {
            Some(q) => q,
            None => return Vec::new(),
        };
        let mut matched = Vec::new();
        let mut remaining = VecDeque::new();
        for token in queue.drain(..) {
            if predicate(token.value.as_ref()) {
                matched.push(token);
            } else {
                remaining.push_back(token);
            }
        }
        *queue = remaining;
        matched
    }

    /// Counts tokens whose values satisfy `predicate`, over the whole queue.
    ///
    /// Public API with no in-crate caller: the ν-net join path (NU-020) used to
    /// size a correlated `all`/`at-least` consume with this, but a same-pass
    /// deposit must stay out of that count ([EXEC-003] AC5), so the backends now
    /// count a bounded prefix inline instead.
    pub fn count_matching(&self, place_name: &str, predicate: &dyn Fn(&dyn Any) -> bool) -> usize {
        self.tokens.get(place_name).map_or(0, |q| {
            q.iter().filter(|t| predicate(t.value.as_ref())).count()
        })
    }

    /// Returns the internal token map (for snapshot/event purposes).
    pub fn token_counts(&self) -> HashMap<Arc<str>, usize> {
        self.tokens
            .iter()
            .filter(|(_, q)| !q.is_empty())
            .map(|(k, q)| (Arc::clone(k), q.len()))
            .collect()
    }

    /// Returns all place names that have tokens.
    pub fn non_empty_places(&self) -> Vec<Arc<str>> {
        self.tokens
            .iter()
            .filter(|(_, q)| !q.is_empty())
            .map(|(k, _)| Arc::clone(k))
            .collect()
    }

    /// Returns the raw queue for a place (for executor internal use).
    pub fn queue(&self, place_name: &str) -> Option<&VecDeque<ErasedToken>> {
        self.tokens.get(place_name)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn add_and_count() {
        let p = Place::<i32>::new("p");
        let mut m = Marking::new();
        assert_eq!(m.count("p"), 0);
        assert!(!m.has_tokens("p"));

        m.add(&p, Token::new(1));
        m.add(&p, Token::new(2));
        assert_eq!(m.count("p"), 2);
        assert!(m.has_tokens("p"));
    }

    #[test]
    fn peek() {
        let p = Place::<i32>::new("p");
        let mut m = Marking::new();
        m.add(&p, Token::new(42));
        assert_eq!(*m.peek(&p).unwrap(), 42);
        assert_eq!(m.count("p"), 1); // peek doesn't consume
    }

    #[test]
    fn remove_first_fifo() {
        let p = Place::<i32>::new("p");
        let mut m = Marking::new();
        m.add(&p, Token::at(1, 100));
        m.add(&p, Token::at(2, 200));

        let t = m.remove_first("p").unwrap();
        let recovered = t.downcast::<i32>().unwrap();
        assert_eq!(*recovered.value(), 1);
        assert_eq!(m.count("p"), 1);
    }

    #[test]
    fn remove_all() {
        let p = Place::<i32>::new("p");
        let mut m = Marking::new();
        m.add(&p, Token::new(1));
        m.add(&p, Token::new(2));
        m.add(&p, Token::new(3));

        let tokens = m.remove_all("p");
        assert_eq!(tokens.len(), 3);
        assert_eq!(m.count("p"), 0);
    }

    #[test]
    fn remove_matching() {
        let p = Place::<i32>::new("p");
        let mut m = Marking::new();
        m.add(&p, Token::new(1));
        m.add(&p, Token::new(2));
        m.add(&p, Token::new(3));

        let predicate = |v: &dyn Any| v.downcast_ref::<i32>().is_some_and(|n| *n > 1);
        let t = m.remove_matching("p", &predicate).unwrap();
        let recovered = t.downcast::<i32>().unwrap();
        assert_eq!(*recovered.value(), 2); // first matching
        assert_eq!(m.count("p"), 2);
    }

    #[test]
    fn token_counts() {
        let p1 = Place::<i32>::new("p1");
        let p2 = Place::<String>::new("p2");
        let mut m = Marking::new();
        m.add(&p1, Token::new(1));
        m.add(&p1, Token::new(2));
        m.add(&p2, Token::new("hello".to_string()));

        let counts = m.token_counts();
        assert_eq!(counts.len(), 2);
        assert_eq!(counts[&Arc::from("p1")], 2);
        assert_eq!(counts[&Arc::from("p2")], 1);
    }
}
